README.md
# Sphinx Build Action (composite action)

[![Build Status](https://travis-ci.com/rickstaa/sphinx-action-composite.svg?branch=master)](https://travis-ci.com/rickstaa/sphinx-action-composite)
[![codecov](https://codecov.io/gh/rickstaa/sphinx-action-composite/branch/master/graph/badge.svg?token=ZCQFPNPWGK)](https://codecov.io/gh/rickstaa/sphinx-action-composite)

This is the [composite variant](https://docs.github.com/en/free-pro-team@latest/actions/creating-actions/creating-a-composite-run-steps-action) of [ammaraskar/sphinx-action](ammaraskar/sphinx-action). This is a Github action that looks for Sphinx documentation folders in your
project. It builds the documentation using Sphinx and any errors in the build
process are bubbled up as Github status checks. This [composite variant](https://docs.github.com/en/free-pro-team@latest/actions/creating-actions/creating-a-composite-run-steps-action) allows you to combine the sphinx-build/linting action with more complex build workflows.

The main purposes of this action are:

* Run a CI test to ensure your documentation still builds.

* Allow contributors to get build errors on simple doc changes inline on Github
  without having to install Sphinx and build locally.

![Example Screenshot](https://i.imgur.com/Gk2W32O.png)

## How to use

Create a workflow for the action, for example:

```yaml
name: "Pull Request Docs Check"
on:
- pull_request

jobs:
  docs:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v1
    - name: Set up Python environment
      uses: actions/setup-python@v2
      with:
        python-version: 3.7
    - name: Update pip
      run: |
        python -m pip install --upgrade pip
    - name: Install the package your documenting together with its dependencies.
      run: |
        pip install .
    - name: Build the sphinx documentation and posts warnings as github comments.
      uses: rickstaa/sphinx-action@master
      with:
        docs-folder: "./docs"
```

* :warning: If you have any Python dependencies that your project needs (themes,
build tools, etc) please be sure you installed these in previous workflow steps. This needs to be done as the the `uses` keyword is not yet available for composite actions (see [actions/runner/issues/646](https://github.com/actions/runner/issues/646)). :warning:

* If you have multiple sphinx documentation folders, please use multiple `uses` blocks.

For a full example repo using this action including advanced usage, take a look
at https://github.com/ammaraskar/sphinx-action-test

## Great Actions to Pair With

Some really good actions that work well with this one are
[`actions/upload-artifact`](https://github.com/actions/upload-artifact)
and [`ad-m/github-push-action`](https://github.com/ad-m/github-push-action).

You can use these to make built HTML and PDFs available as artifacts:

```yaml
    - uses: actions/upload-artifact@v1
      with:
        name: DocumentationHTML
        path: docs/_build/html/
```

Or to push docs changes automatically to a `gh-pages` branch:

<details><summary>Code for your workflow</summary>
<p>

```yaml
    - name: Commit documentation changes
      run: |
        git clone https://github.com/ammaraskar/sphinx-action-test.git --branch gh-pages --single-branch gh-pages
        cp -r docs/_build/html/* gh-pages/
        cd gh-pages
        git config --local user.email "action@github.com"
        git config --local user.name "GitHub Action"
        git add .
        git commit -m "Update documentation" -a || true
        # The above command will fail if no changes were present, so we ignore
        # the return code.
    - name: Push changes
      uses: ad-m/github-push-action@master
      with:
        branch: gh-pages
        directory: gh-pages
        github_token: ${{ secrets.GITHUB_TOKEN }}
```

</p>
</details>

For a full fledged example of this in action take a look at:
https://github.com/ammaraskar/sphinx-action-test

## Advanced Usage

If you wish to customize the command used to build the docs (defaults to
`make html`), you can provide a `build-command` in the `with` block. For
example, to invoke sphinx-build directly you can use:

```yaml
    - uses: ammaraskar/sphinx-action@master
      with:
        docs-folder: "docs/"
        build-command: "sphinx-build -b html . _build"
```

If there's system level dependencies that need to be installed for your
build, you have to install them in the previous workflow steps.

## Running the tests

`python -m unittest`

## Formatting

Please use [black](https://github.com/psf/black) for formatting:

`black entrypoint.py sphinx_action tests`


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v1.0](https://github.com/rickstaa/sphinx-action-composite/compare/v1...v1.0)

> 23 December 2020

#### [v1](https://github.com/rickstaa/sphinx-action-composite/compare/v1.0.0...v1)

> 23 December 2020

### [v1.0.0](https://github.com/rickstaa/sphinx-action-composite/compare/0.4...v1.0.0)

> 23 December 2020

- Change Dockerfile to use sphinxdoc/sphinx base image [`#12`](https://github.com/rickstaa/sphinx-action-composite/pull/12)
- :sparkles: Replaces the docker action with a composite action [`ebd7de6`](https://github.com/rickstaa/sphinx-action-composite/commit/ebd7de69b3402945116f0cb81479f11c91c9c2f9)
- :construction_worker: Adds action release gh-action [`b00a93d`](https://github.com/rickstaa/sphinx-action-composite/commit/b00a93da773f9591f7f26287102017a3ee4acfd7)
- :bug: Makes sure the action inputs are piped to the sphinx build/lint script [`19889fc`](https://github.com/rickstaa/sphinx-action-composite/commit/19889fc36291858903c72e87849f9b360befc236)

#### [0.4](https://github.com/rickstaa/sphinx-action-composite/compare/0.3...0.4)

> 15 February 2020

- Document how to upload artifacts and push gh-pages. Fixes #8 [`#8`](https://github.com/rickstaa/sphinx-action-composite/issues/8)
- Use black for formatting [`cbec8bb`](https://github.com/rickstaa/sphinx-action-composite/commit/cbec8bb26e2580fa17151951c0b50ce4d2773853)
- Switch to a logging based approach instead of check-runs API [`d13c86e`](https://github.com/rickstaa/sphinx-action-composite/commit/d13c86e3e3e6ccee46cdf8023e977694a8692e70)
- Pass -e into `make` to force environment overrides [`1983868`](https://github.com/rickstaa/sphinx-action-composite/commit/1983868b92093bed61a91ac46eaa2d173d897872)

#### [0.3](https://github.com/rickstaa/sphinx-action-composite/compare/0.2...0.3)

> 23 December 2019

- Pass docs folder in inputs to make interface a little nicer [`97242e8`](https://github.com/rickstaa/sphinx-action-composite/commit/97242e8da9df6de528b8923bb331cafe9f76e08b)
- Remove unused sys import [`3dd694e`](https://github.com/rickstaa/sphinx-action-composite/commit/3dd694e9195d6967845dd6d9832f46d879aeb790)

#### [0.2](https://github.com/rickstaa/sphinx-action-composite/compare/0.1...0.2)

> 23 December 2019

- Allow customizing build command [`54a8d91`](https://github.com/rickstaa/sphinx-action-composite/commit/54a8d91610ea50c75ef245442418994a35705832)
- Add info on new features to README [`2e99655`](https://github.com/rickstaa/sphinx-action-composite/commit/2e996556da98aa8189f221d1b283de086a8f05f4)
- Allow running a command before the action [`bbd4951`](https://github.com/rickstaa/sphinx-action-composite/commit/bbd49514d9b0cce3028e70e1fc970c0d181419cc)

#### 0.1

> 31 August 2019

- Add more test projects [`476ca8e`](https://github.com/rickstaa/sphinx-action-composite/commit/476ca8e77b4ea851ad8756503979e6d3b2f0c1c1)
- Initial commit [`1f868f2`](https://github.com/rickstaa/sphinx-action-composite/commit/1f868f243a4062863f2c52ea1d743ee91a40a45b)
- Finish out log parsing code and tests [`09e63d0`](https://github.com/rickstaa/sphinx-action-composite/commit/09e63d05db424a029e6a1914d6105d1f48aad0fd)


tests/test_action.py
import os
import unittest
from unittest import mock

from sphinx_action import action, status_check


TEST_PROJECTS_DIR = os.path.join(
    os.path.dirname(os.path.realpath(__file__)), "test_projects"
)


class TestAction(unittest.TestCase):
    def test_extract_line_information_with_no_line_info(self):
        self.assertEqual(
            action.extract_line_information("warnings_and_errors/index.rst: "),
            ("warnings_and_errors{}index.rst".format(os.sep), 1),
        )

    def test_extract_line_information_with_line_info(self):
        self.assertEqual(
            action.extract_line_information("warnings_and_errors/contents.rst:19: "),
            ("warnings_and_errors{}contents.rst".format(os.sep), 19),
        )

    def test_extract_line_information_with_invalid_info(self):
        self.assertEqual(action.extract_line_information("adsfafadf"), None)

    def test_extract_line_information_with_invalid_lineno(self):
        self.assertEqual(action.extract_line_information("index.rst:a:"), None)

    def test_build_docs_with_no_docs_directories(self):
        """Make sure we bail out if they don't provide a single documentation
        directory"""
        with self.assertRaises(ValueError):
            action.build_all_docs(None, [])

    def test_build_docs_no_errors_or_warnings(self):
        """Check that we correctly build docs when there's no errors or
        warnings"""
        return_code, annotations = action.build_docs(
            "make html", os.path.join(TEST_PROJECTS_DIR, "no_errors")
        )
        self.assertEqual(return_code, 0)
        self.assertEqual(annotations, [])

    def test_build_with_custom_command(self):
        return_code, annotations = action.build_docs(
            "sphinx-build -b html . _build",
            os.path.join(TEST_PROJECTS_DIR, "no_errors"),
        )
        self.assertEqual(return_code, 0)
        self.assertEqual(annotations, [])

    def test_build_docs_errors(self):
        return_code, annotations = action.build_docs(
            "make html", os.path.join(TEST_PROJECTS_DIR, "errors")
        )
        self.assertEqual(annotations, [])

    def test_build_docs_warnings(self):
        return_code, annotations = action.build_docs(
            "make html", os.path.join(TEST_PROJECTS_DIR, "warnings")
        )
        self.assertEqual(return_code, 0)
        self.assertEqual(len(annotations), 3)

        self.assertTrue(annotations[0].path.endswith("index.rst"))
        self.assertEqual(annotations[0].start_line, 22)
        self.assertEqual(annotations[0].end_line, 22)
        self.assertIn('Problems with "include"', annotations[0].message)

        self.assertTrue(annotations[1].path.endswith("index.rst"))
        self.assertEqual(annotations[1].start_line, 24)
        self.assertEqual(annotations[1].end_line, 24)
        self.assertEqual(
            'Unknown directive type "BADDIRECTIVE".', annotations[1].message
        )

        self.assertTrue(annotations[2].path.endswith("notintoc.rst"))
        self.assertEqual(annotations[2].start_line, 1)
        self.assertEqual(annotations[2].end_line, 1)
        self.assertEqual(
            "document isn't included in any toctree", annotations[2].message
        )

    def test_build_docs_warnings_and_errors(self):
        return_code, annotations = action.build_docs(
            "make html", os.path.join(TEST_PROJECTS_DIR, "warnings_and_errors")
        )
        self.assertNotEqual(return_code, 0)
        self.assertEqual(len(annotations), 1)

        self.assertTrue(annotations[0].path.endswith("index.rst"))
        self.assertEqual(annotations[0].start_line, 16)
        self.assertEqual(annotations[0].end_line, 16)
        self.assertIn('Error in "code-block" directive', annotations[0].message)

    def test_build_docs_with_different_makefile(self):
        """Test some different styles of sphinx Makefiles that don't
        use ?= for the options"""
        return_code, annotations = action.build_docs(
            "make html", os.path.join(TEST_PROJECTS_DIR, "different_makefile")
        )
        self.assertNotEqual(return_code, 0)
        self.assertEqual(len(annotations), 1)

        self.assertTrue(annotations[0].path.endswith("index.rst"))
        self.assertEqual(annotations[0].start_line, 16)
        self.assertEqual(annotations[0].end_line, 16)
        self.assertIn('Error in "code-block" directive', annotations[0].message)

    def test_build_all_docs_success(self):
        action.build_all_docs(
            action.GithubEnvironment(
                build_command="make html",
            ),
            [
                os.path.join(TEST_PROJECTS_DIR, "no_errors"),
                os.path.join(TEST_PROJECTS_DIR, "warnings"),
            ],
        )

    def test_build_all_docs_some_success(self):
        with self.assertRaisesRegex(RuntimeError, "Build failed"):
            action.build_all_docs(
                action.GithubEnvironment(
                    build_command="make html",
                ),
                [
                    os.path.join(TEST_PROJECTS_DIR, "no_errors"),
                    os.path.join(TEST_PROJECTS_DIR, "warnings_and_errors"),
                ],
            )


if __name__ == "__main__":
    unittest.main()


sphinx_action/action.py
import collections
import subprocess
import tempfile
import os
import shlex

from sphinx_action import status_check


GithubEnvironment = collections.namedtuple("GithubEnvironment", ["build_command"])


def extract_line_information(line_information):
    r"""Lines from sphinx log files look like this

        C:\Users\ammar\workspace\sphinx-action\tests\test_projects\warnings\index.rst:22: WARNING: Problems with "include" directive path:
        InputError: [Errno 2] No such file or directory: 'I_DONT_EXIST'.

        /home/users/ammar/workspace/sphix-action/tests/test_projects/warnings/index.rst:22: WARNING: Problems with "include" directive path:
        InputError: [Errno 2] No such file or directory: 'I_DONT_EXIST'.

        /home/users/ammar/workspace/sphix-action/tests/test_projects/warnings/index.rst: Something went wrong with this whole ifle

    This method is responsible for parsing out the line number and file name from these lines.
    """
    file_and_line = line_information.split(":")
    # This is a dirty windows specific hack to deal with drive letters in the
    # start of the file-path, i.e D:\
    if len(file_and_line[0]) == 1:
        # If the first component is just one letter, we did an accidental split
        file_and_line[1] = file_and_line[0] + ":" + file_and_line[1]
        # Join the first component back up with the second and discard it.
        file_and_line = file_and_line[1:]

    if len(file_and_line) != 2 and len(file_and_line) != 3:
        return None
    # The case where we have no line number, in this case we return the line
    # number as 1 to mark the whole file.
    if len(file_and_line) == 2:
        line_num = 1
    if len(file_and_line) == 3:
        try:
            line_num = int(file_and_line[1])
        except ValueError:
            return None

    file_name = os.path.relpath(file_and_line[0])
    return file_name, line_num


def parse_sphinx_warnings_log(logs):
    """Parses a sphinx file containing warnings and errors into a list of
        status_check.CheckAnnotation objects.

        Inputs look like this:
    /media/sf_shared/workspace/sphinx-action/tests/test_projects/warnings_and_errors/index.rst:19: WARNING: Error in "code-block" directive:
    maximum 1 argument(s) allowed, 2 supplied.

    /cpython/Doc/distutils/_setuptools_disclaimer.rst: WARNING: document isn't included in any toctree
    /cpython/Doc/contents.rst:5: WARNING: toctree contains reference to nonexisting document 'ayylmao'
    """
    annotations = []

    for i, line in enumerate(logs):
        if "WARNING" not in line:
            continue

        warning_tokens = line.split("WARNING:")
        if len(warning_tokens) != 2:
            continue
        file_and_line, message = warning_tokens

        file_and_line = extract_line_information(file_and_line)
        if not file_and_line:
            continue
        file_name, line_number = file_and_line

        warning_message = message
        # If this isn't the last line and the next line isn't a warning,
        # treat it as part of this warning message.
        if (i != len(logs) - 1) and "WARNING" not in logs[i + 1]:
            warning_message += logs[i + 1]
        warning_message = warning_message.strip()

        annotations.append(
            status_check.CheckAnnotation(
                path=file_name,
                message=warning_message,
                start_line=line_number,
                end_line=line_number,
                annotation_level=status_check.AnnotationLevel.WARNING,
            )
        )

    return annotations


def build_docs(build_command, docs_directory):
    if not build_command:
        raise ValueError("Build command may not be empty")

    log_file = os.path.join(tempfile.gettempdir(), "sphinx-log")
    if os.path.exists(log_file):
        os.unlink(log_file)

    sphinx_options = '--keep-going --no-color -w "{}"'.format(log_file)
    # If we're using make, pass the options as part of the SPHINXOPTS
    # environment variable, otherwise pass them straight into the command.
    build_command = shlex.split(build_command)
    if build_command[0] == "make":
        # Pass the -e option into `make`, this is specified to be
        #   Cause environment variables, including those with null values, to override macro assignments within makefiles.
        # which is exactly what we want.
        build_command += ["-e"]
        print("[sphinx-action] Running: {}".format(build_command))

        return_code = subprocess.call(
            build_command,
            env=dict(os.environ, SPHINXOPTS=sphinx_options),
            cwd=docs_directory,
        )
    else:
        build_command += shlex.split(sphinx_options)
        print("[sphinx-action] Running: {}".format(build_command))

        return_code = subprocess.call(
            build_command + shlex.split(sphinx_options), cwd=docs_directory
        )

    with open(log_file, "r") as f:
        annotations = parse_sphinx_warnings_log(f.readlines())

    return return_code, annotations


def build_all_docs(github_env, docs_directories):
    if len(docs_directories) == 0:
        raise ValueError("Please provide at least one docs directory to build")

    build_success = True
    warnings = 0

    for docs_dir in docs_directories:
        print("====================================")
        print("Building docs in {}".format(docs_dir))
        print("====================================")

        return_code, annotations = build_docs(github_env.build_command, docs_dir)
        if return_code != 0:
            build_success = False

        warnings += len(annotations)

        for annotation in annotations:
            status_check.output_annotation(annotation)

    status_message = "[sphinx-action] Build {} with {} warnings".format(
        "succeeded" if build_success else "failed", warnings
    )
    print(status_message)

    if not build_success:
        raise RuntimeError("Build failed")
